File: utils/py/singly_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def to_list(self):
        r = []
        while self is not None:
            r.append(self.val)
            self = self.next
        return r
    @staticmethod
    def from_list(li):
        head = ListNode()
        cur = head
        for e in li:
            cur.next = ListNode(e)
            cur = cur.next
        return head.next
    
if __name__ == '__main__':
    li = ListNode.from_list([1,2,3,4])
    print(li.to_list())

File: utils/golang/singly_linked_list.go
package golang

type ListNode[E any] struct {
	Val  E
	Next *ListNode[E]
}

func (l *ListNode[E]) ToSlice() []E {
	var r []E
	for l != nil {
		r = append(r, l.Val)
		l = l.Next
	}
	return r
}

func MakeSinglyLinkedList[E any](es ...E) *ListNode[E] {
	head := ListNode[E]{}
	cur := &head
	for _, v := range es {
		cur.Next = &ListNode[E]{Val: v}
		cur = cur.Next
	}

	return head.Next
}
